github/codeowners.py

```diff
@@ -116,38 +118,146 @@ def resolve_team_members(
     if not team_or_none:
         logger.warning('failed to lookup team {t}'.format(t=team_name))
         return []
-    for member in map(github_api.user, team_or_none.members()):
-        if member.email:
-            yield member.email
+    yield from map(github_api.user, team_or_none.members())
+
+
+@dataclasses.dataclass
+class CodeownerEntry:
+    value: str
+
+
+@dataclasses.dataclass
+class UsernameCodeownerEntry(CodeownerEntry):
+    pass
+
+
+@dataclasses.dataclass
+class EmailCodeownerEntry(CodeownerEntry):
+    pass
+
+
+@dataclasses.dataclass
+class TeamnameCodeownerEntry(CodeownerEntry):
+    pass
+
+
+class Filter:
+    @staticmethod
+    def usernames(codeowner_entry: CodeownerEntry) -> bool:
+        return isinstance(codeowner_entry, UsernameCodeownerEntry)
+
+    @staticmethod
+    def emails(codeowner_entry: CodeownerEntry) -> bool:
+        return isinstance(codeowner_entry, EmailCodeownerEntry)
+
+    @staticmethod
+    def teamnames(codeowner_entry: CodeownerEntry) -> bool:
+        return isinstance(codeowner_entry, TeamnameCodeownerEntry)
+
+
+def iter_codeowners(
+    codeowners_entries,
+    node_filter: typing.Callable[[CodeownerEntry], bool]=None,
+) -> typing.Generator[str, None, None]:
+    '''
+    returns a generator yielding the codeowner entries.
+
+    @param codeowner_entries:  root for iteration
+    @node_filter:              use to filter emitted nodes (see Filter for predefined filters)
```

Comment:
you never call the emitted objects nodes, neither in code, not in doc. why call them nodes here? rather name this just filter, or entry_filter.